packages/flutter/lib/src/widgets/focus_manager.dart

```diff
@@ -3,50 +3,343 @@
 // found in the LICENSE file.
 
 import 'dart:async';
+import 'dart:ui';
 
 import 'package:flutter/foundation.dart';
+import 'package:flutter/painting.dart';
+import 'package:flutter/services.dart';
 
-/// A leaf node in the focus tree that can receive focus.
+import 'binding.dart';
+import 'framework.dart';
+
+/// Signature of a callback used by [Focus.onKey] and [FocusScope.onKey]
+/// to receive key events.
 ///
-/// The focus tree keeps track of which widget is the user's current focus. The
-/// focused widget often listens for keyboard events.
+/// The [node] is the node that received the event.
+typedef FocusOnKeyCallback = bool Function(FocusNode node, RawKeyEvent event);
+
+/// An attachment point for a [FocusNode].
 ///
-/// To request focus, find the [FocusScopeNode] for the current [BuildContext]
-/// and call the [FocusScopeNode.requestFocus] method:
+/// Once created, a [FocusNode] must be attached to the focus tree via a
+/// [FocusAttachment] object.  This attachment is created by calling
+/// [FocusNode.attach], usually from a widget's [State.initState] method, and if
+/// the widget is updated, again in [State.didUpdateWidget], after calling
```

Comment:
why do you need to do it in didUpdateWidget? If you attach to it in initState, I would expect that to just require a detach in dispose and the reparent in build.